sysl2/sysl/deps.go

```diff
@@ -0,0 +1,284 @@
+package main
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/anz-bank/sysl/src/proto"
+	"github.com/anz-bank/sysl/sysl2/sysl/utils"
+	"github.com/yale8848/stream"
+)
+
+type AppElement struct {
+	Name     string
+	Endpoint string
+}
+
+type AppDependency struct {
+	Self   *AppElement
+	Target *AppElement
+}
+
+type DependencySet struct {
+	Deps []*AppDependency
+}
+
+func (dep *AppDependency) String() string {
+	return fmt.Sprintf("s.name: %s, s.end: %s, t.name: %s, t.end: %s", dep.Self.Name, dep.Self.Endpoint, dep.Target.Name, dep.Target.Endpoint)
+}
+
+func (dep *AppDependency) Equals(target *AppDependency) bool {
+	return (dep.Self.Name == target.Self.Name) && (dep.Self.Endpoint == target.Self.Endpoint) &&
+		(dep.Target.Name == target.Target.Name) && (dep.Target.Endpoint == target.Target.Endpoint)
+}
+
+func (ds *DependencySet) Add(dep *AppDependency) {
+	if !ds.Contains(dep) {
+		ds.Deps = append(ds.Deps, dep)
+	}
+}
+
+func (ds *DependencySet) Contains(dep *AppDependency) bool {
+	for _, v := range ds.Deps {
+		if v.Equals(dep) {
+			return true
+		}
```

Comment:
Contains is inefficient. if call list was having small number of items its ok. but for a large project, this will start to hurt.

Reply:
change slice structure to map and no need for this function.